openquake/risklib/reinsurance.py

```diff
@@ -41,14 +42,72 @@
 DEBUG = False
 
 
-def check_fields(fields, dframe, idxdict, fname):
+def check_fields(fields, dframe, idxdict, fname, policyfname, treaties,
+                 treaty_types):
     """
     :param fields: fields to check (the first field is the primary key)
     :param dframe: DataFrame with the contents of fname
     :param idxdict: dictionary key -> index (starting from 1)
-    :param fname: file containing the fields to check
+    :param fname: file xml containing the fields to check
+    :param policyfname: file csv containing the fields to check
+    :param treaties: treaty names
+    :param treaty_types: treaty types
     """
     key = fields[0]
+    if len(dframe[key].unique()) < len(dframe):
+        raise InvalidFile(f'{policyfname}: {key} contains duplicates')
+    for treaty in treaties:
+        if treaty not in dframe.columns:
+            raise InvalidFile(f'{policyfname}: {treaty} is missing')
+    policies_from_exposure = list(idxdict)[1:]  # discard '?'
+    policies_from_csv = list(dframe.policy)
+    for policy in policies_from_exposure:
+        if policy not in policies_from_csv:
```

Comment:
This is really slow; change it to use numpy.isin or something like that.

Reply:
Done